src/ModelContextProtocol/Shared/McpSession.cs

```diff
@@ -296,6 +296,18 @@ await _transport.SendMessageAsync(new JsonRpcResponse
         }, cancellationToken).ConfigureAwait(false);
     }
 
+    private void RegisterCancellation(CancellationToken cancellationToken, RequestId requestId)
+    {
+        cancellationToken.Register(async () => await SendMessageAsync(new JsonRpcNotification
```

Comment:
It's important to Dispose of the CancellationTokenRegistration. This RegisterCancellation method should return that CancellationTokenRegistration, and the call site should use it in a using around the rest of the operation.